server/spec/services/mongo_pubsub_spec.rb

```diff
@@ -96,15 +96,20 @@
       sub1.terminate
       sub2.terminate
 
-      thread_count = Thread.list.count
+      described_class.publish('test', nil)
+      GC.start
+
+      thread_list = Thread.list
 
       sub1 = described_class.subscribe('channel1') {|msg| }
       sub2 = described_class.subscribe('channel1') {|msg| }
       sub1.terminate
       sub2.terminate
 
+      described_class.publish('test', nil)
       GC.start
-      expect(thread_count == Thread.list.count).to be_truthy
+
+      expect(Thread.list).to eq thread_list
```

Comment:
Nope... still possible for this to fail after a large number of tries... the sudden appearance of the `dead` threads here is pretty mysterious to me.

```
  1) MongoPubsub.publish cleanups threads
     Failure/Error: expect(Thread.list).to eq thread_list
       
       expected: [#<Thread:0x000000015ca1b0 run>, #<Celluloid::Thread:0x00000002a59608@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a52a88@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a4af68@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a45310@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x000000047f38d0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.3/lib/mongo/server/monitor.rb:121 sleep>, #<Thread:0x000000047f3600@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.3/lib/mongo/cluster/cursor_reaper.rb:166 sleep>, #<Celluloid::Thread:0x00000004a17e18@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000004a14f10@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000004a03620@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x000000049eba98@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x000000049e88c0@/home/kontena/kontena/kontena/server/app/initializers/eventmachine.rb:8 sleep>, #<Celluloid::Thread:0x00000006024d50@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x007fa2d0026460@/usr/lib/ruby/2.3.0/timeout.rb:81 sleep>, #<Celluloid::Thread:0x00000002e12038@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d0050f58@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d005bde0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d0053078@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2c8071468@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>]
            got: [#<Thread:0x000000015ca1b0 run>, #<Celluloid::Thread:0x00000002a59608@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a52a88@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a4af68@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000002a45310@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x000000047f38d0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.3/lib/mongo/server/monitor.rb:121 sleep>, #<Thread:0x000000047f3600@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.3/lib/mongo/cluster/cursor_reaper.rb:166 sleep>, #<Celluloid::Thread:0x00000004a17e18@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000004a14f10@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x00000004a03620@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x000000049eba98@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x000000049e88c0@/home/kontena/kontena/kontena/server/app/initializers/eventmachine.rb:8 sleep>, #<Celluloid::Thread:0x00000006024d50@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Thread:0x007fa2d0026460@/usr/lib/ruby/2.3.0/timeout.rb:81 sleep>, #<Celluloid::Thread:0x00000002e12038@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d0050f58@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d005bde0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2d0053078@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x007fa2c8071468@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>, #<Celluloid::Thread:0x000000017ad220@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 dead>]
       
       (compared using ==)
       
       Diff:
       @@ -16,5 +16,6 @@
         #<Celluloid::Thread:0x007fa2d0050f58@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007fa2d005bde0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007fa2d0053078@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       - #<Celluloid::Thread:0x007fa2c8071468@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>]
       + #<Celluloid::Thread:0x007fa2c8071468@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       + #<Celluloid::Thread:0x000000017ad220@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 dead>]
       
     # ./spec/services/mongo_pubsub_spec.rb:112:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:100:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:93:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:92:in `block (2 levels) in <top (required)>'
```